### PR TITLE
`Tutorial groups`: Allow students to view tutorial groups and tutors to register students

### PR DESCRIPTION
**Pull request by stefanwaldhauser**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

The following user stories are implemented:
- As a student I want to check to which tutorial groups I am registered, so I know when and where I have to attend.
- As a student I want to see all tutorial groups, so I possibly can contact another tutor to change my tutorial group.
- As a tutor of a tutorial group, I want to register students to my group, so I do not have to contact the instructor for this every time.
- As a tutor of a tutorial group, I want to derigster students from my group, so I do not have to contact the instructor for this every time.

### Description

_Changes to Client Code_:
- Implemented the course tutorial group overview and detail views with as little code duplication as possible by making previous components generic and reusable.
- Allowed tutors to also access tutorial group management with managing registrations for their own groups as the only functionality.

_Changes to Server Code_:
- Made the `getOne` and `getAllOfCourse` endpoints available for students and tutors. Privacy sensitive information like the actual registrations of students is removed before sending the information to the client if the user is not at least the tutor of the tutorial group.
- Made the `register` and `deregister` endpoints available for the tutor of a tutorial group.
- Added transient properties to the `TutorialGroup` entity, to tell the client if a user is registered or the assigned tutor to a respective tutorial group and the total number of registered students.

_Changes Related to Client Testing:_
- Vastly improved client test setup and coverage

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise with Complaints enabled

1. Log in to Artemis as an Administrator
2. Activate the Tutorial Group Feature Toggle in the admin settings
3. Create Multiple Tutorial Groups and register some students for them
4. Login as one of the registered students and confirm that you see your registered tutorial group and all othe tutorial groups (see screenshots)

#### Code Review
- [x] Review 1
- [ ] Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### Test Coverage

Basically all test changes on the client

| Class/File | Branch | Line |
|------------|-------:|-----:|
| course-tutorial-groups.component.ts | 81% | 89% |
| course-tutorial-group-card.component.ts | 90% | 100% |
| course-tutorial-groups-overview.component.ts | 75% | 100% |
| course-tutorial-groups-registered.component.ts| 100% | 100% |
| tutorial-group-detail.component.ts | 100% | 100% |
| tutorial-groups-table.component.ts | 100% | 100% |

### Screenshots

**Student View**

image
image
image
image
image
image

**Tutor View**
image
image
image

### CODE REVIEW

File: src/main/webapp/app/course/tutorial-groups/services/tutorial-group-registration.service.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { HttpClient, HttpResponse } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { TutorialGroupRegistration } from 'app/entities/tutorial-group/tutorial-group-registration.model';
+import { Observable } from 'rxjs';
+type EntityResponseType = HttpResponse<TutorialGroupRegistration>;
```
Comment (JohannesStoehr) [inline src/main/webapp/app/course/tutorial-groups/services/tutorial-group-registration.service.ts:5]: Codacy

File: src/main/webapp/app/course/tutorial-groups/tutorial-groups-instructor-view/tutorial-groups/crud/create-tutorial-group/create-tutorial-group.component.ts [hunk 2]
```diff
@@ -27,11 +27,6 @@ export class CreateTutorialGroupComponent implements OnInit {
     }
 
     createTutorialGroup(formData: TutorialGroupFormData) {
-        // required fields
-        if (!formData?.title || !formData?.teachingAssistant) {
-            return;
-        }
```
Comment (JohannesStoehr) [inline src/main/webapp/app/course/tutorial-groups/tutorial-groups-instructor-view/tutorial-groups/crud/create-tutorial-group/create-tutorial-group.component.ts:33]: Is this no longer required?

File: src/main/webapp/i18n/de/courseTutorialGroups.json [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        },
+        "pages": {
+            "courseTutorialGroups": {
+                "title": "Deine angemeldeten Übungsgruppen",
+                "noRegistrations": "Du bist in keiner Übungsgruppe angemeldet.",
+                "allFilter": "Alle Übungsgruppen anzeigen",
+                "registeredFilter": "Angemeldete Übungsgruppen anzeigen"
+            },
+            "courseTutorialGroupDetail": {
+                "title": "Übungsgruppen Details"
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/courseTutorialGroups.json:14]: Either hyphenate or make one word

File: src/main/webapp/i18n/de/courseTutorialGroups.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            },
+            "courseTutorialGroupOverview": {
+                "title": "Übungsgruppen Übersicht",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/courseTutorialGroups.json:17]: Either hyphenate or make one word or "Übersicht der Übungsgruppen"

File: src/main/webapp/i18n/en/courseTutorialGroups.json [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        },
+        "pages": {
+            "courseTutorialGroups": {
+                "title": "Your registered Tutorial Groups",
+                "noRegistrations": "You are not registered to any tutorial group in this course.",
+                "allFilter": "Show all Tutorial Groups",
+                "registeredFilter": "Show registered Tutorial Groups"
+            },
+            "courseTutorialGroupDetail": {
+                "title": "Tutorial Group Detail"
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/en/courseTutorialGroups.json:14]: In the German version it's DetailS

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

**Review by manuelmanso — CHANGES_REQUESTED** [review]
Tested in ts1, works nicely but some changes are required as discussed in the testing session:

- "Show tutorial groups" button is clipping
- Always show the "Course information"
- Hide the teaching assisstant's information (other than name) when fetching tutorial groups
- Hide the other student's information when fetching registrations (only missing in for-dashboard call)

File: src/main/webapp/app/course/tutorial-groups/tutorial-groups-instructor-view/tutorial-groups/crud/create-tutorial-group/create-tutorial-group.component.ts (hunk 2 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/course/tutorial-groups/tutorial-groups-instructor-view/tutorial-groups/crud/create-tutorial-group/create-tutorial-group.component.ts:33]: The form submission is disabled if the required fields are missing, so this is code duplication basically. Therefore I removed it

File: src/main/webapp/i18n/de/courseTutorialGroups.json (hunk 4 above)
Comment (stefanwaldhauser) [reply src/main/webapp/i18n/de/courseTutorialGroups.json]: Done in 1a8ae32df3469f96ada97d6dfa074467d675a5ec

**Comment by stefanwaldhauser** [comment]
> Tested in ts1, works nicely but some changes are required as discussed in the testing session:
> 
> * "Show tutorial groups" button is clipping
> * Always show the "Course information"
> * Hide the teaching assisstant's information (other than name) when fetching tutorial groups
> * Hide the other student's information when fetching registrations (only missing in for-dashboard call)

I addressed everything except the clipping. Its a more fundamental issue. I opened a bug report here: https://github.com/ls1intum/Artemis/issues/5724

File: src/test/javascript/spec/component/tutorial-groups/course-tutorial-groups/course-tutorial-groups-overview.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+import { CourseTutorialGroupsOverviewComponent } from 'app/overview/course-tutorial-groups/course-tutorial-groups-overview/course-tutorial-groups-overview.component';
+import { Component, Input } from '@angular/core';
```
Comment (ge65cer) [inline src/test/javascript/spec/component/tutorial-groups/course-tutorial-groups/course-tutorial-groups-overview.component.spec.ts:3]: Unused import.

File: src/test/javascript/spec/component/tutorial-groups/helpers/tutorialGroupExampleModels.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    language = Language.ENGLISH,
+    additionalInformation = 'Example Additional Information',
+    isOnline = false,
+    teachingAssistant = { id: 1, login: 'Example TA' } as User,
+    isUserRegistered = false,
+    numberOfRegisteredUsers = 5,
+    course = { id: 1, title: 'Test Course' } as Course,
+    teachingAssistantName = 'Example TA',
+    courseTitle = 'Test Course',
+}: GenerateExampleTutorialGroupParams) => {
```
Comment (ge65cer) [inline src/test/javascript/spec/component/tutorial-groups/helpers/tutorialGroupExampleModels.ts:36]: Why do you not use `as TutorialGroup`?

File: src/test/javascript/spec/component/tutorial-groups/course-tutorial-groups/course-tutorial-group-detail.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    let component: CourseTutorialGroupDetailComponent;
+
+    beforeEach(() => {
+        TestBed.configureTestingModule({
+            declarations: [CourseTutorialGroupDetailComponent, TutorialGroupDetailStubComponent, LoadingIndicatorContainerStubComponent],
+            providers: [
+                MockProvider(TutorialGroupsService),
+                MockProvider(AlertService),
+                { provide: Router, useClass: MockRouter },
+                simpleOneLayerActivatedRouteProvider(
```
Comment (ge65cer) [inline src/test/javascript/spec/component/tutorial-groups/course-tutorial-groups/course-tutorial-group-detail.component.spec.ts:26]: Do you know about `convertToParamMap()`?

File: src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java [hunk 9]
```diff
@@ -62,6 +62,24 @@ public class TutorialGroup extends DomainObject {
...
+    @JsonSerialize
+    private Integer numberOfRegisteredUsers;
+
+    @Transient
+    @JsonSerialize
+    private String teachingAssistantName;
+
+    @Transient
+    @JsonSerialize
+    private String courseTitle;
```
Comment (ge65cer) [inline src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java:81]: When is this needed? Can you not re-use the course object in the client?

File: src/main/webapp/app/course/tutorial-groups/tutorial-groups-instructor-view/tutorial-groups/tutorial-groups-management/tutorial-groups-management.component.html [hunk 10]
```diff
@@ -10,77 +10,14 @@ <h4>{{ 'artemisApp.pages.tutorialGroupsManagement.title' | artemisTranslate }}</
...
-                                    [courseId]="courseId"
-                                    [tutorialGroup]="tutorialGroup"
-                                    (tutorialGroupDeleted)="loadTutorialGroups()"
-                                ></jhi-tutorial-group-row-buttons>
-                            </td>
-                        </tr>
-                    </tbody>
-                </table>
-            </div>
+            <ng-container *ngIf="tutorialGroups && tutorialGroups.length > 0">
```
Comment (ge65cer) [inline src/main/webapp/app/course/tutorial-groups/tutorial-groups-instructor-view/tutorial-groups/tutorial-groups-management/tutorial-groups-management.component.html:13]: Please test, this should work too. 🙂 
```suggestion
            <ng-container *ngIf="tutorialGroups?.length">
```

File: src/main/webapp/app/overview/course-overview.component.ts [hunk 11]
```diff
@@ -192,12 +200,17 @@ export class CourseOverviewComponent implements OnInit, OnDestroy, AfterViewInit
...
     }
 
-    loadLearningGoals() {
-        forkJoin([this.learningGoalService.getAllForCourse(this.courseId), this.learningGoalService.getAllPrerequisitesForCourse(this.courseId)]).subscribe({
-            next: ([learningGoals, prerequisites]) => {
+    loadLearningGoalsAndTutorialGroups() {
+        forkJoin([
+            this.learningGoalService.getAllForCourse(this.courseId),
+            this.learningGoalService.getAllPrerequisitesForCourse(this.courseId),
+            this.tutorialGroupService.getAllOfCourse(this.courseId),
```
Comment (ge65cer) [inline src/main/webapp/app/overview/course-overview.component.ts:207]: Maybe make the method naming consistent with the other functions?
```suggestion
            this.tutorialGroupService.getAllForCourse(this.courseId),
```

**Review by ge65cer — COMMENTED** [review]
Left some comments. 🙂 I don't know how helpful the side panel (number of registrations) is for students? This should probably be only on the course management page.
Screenshot 2022-10-07 at 21 19 14

File: src/test/javascript/spec/component/tutorial-groups/course-tutorial-groups/course-tutorial-groups-overview.component.spec.ts (hunk 6 above)
Comment (stefanwaldhauser) [reply src/test/javascript/spec/component/tutorial-groups/course-tutorial-groups/course-tutorial-groups-overview.component.spec.ts]: Removed in 404a61f9e625e0e40fb810485d800baaea28b685

File: src/test/javascript/spec/component/tutorial-groups/helpers/tutorialGroupExampleModels.ts (hunk 7 above)
Comment (stefanwaldhauser) [reply src/test/javascript/spec/component/tutorial-groups/helpers/tutorialGroupExampleModels.ts]: Good idea! Changed in 404a61f9e625e0e40fb810485d800baaea28b685

File: src/test/javascript/spec/component/tutorial-groups/course-tutorial-groups/course-tutorial-group-detail.component.spec.ts (hunk 8 above)
Comment (stefanwaldhauser) [reply src/test/javascript/spec/component/tutorial-groups/course-tutorial-groups/course-tutorial-group-detail.component.spec.ts]: Good idea! Removed the simple implementation in 404a61f9e625e0e40fb810485d800baaea28b685

File: src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java (hunk 9 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java:81]: I like it because it makes certain parts much cleaner. For example then I do not have to pass the course separately as an extra `@Input` to sub components just for displaying the course title. Is keeping it in a hard blocker for you?

File: src/main/webapp/app/course/tutorial-groups/tutorial-groups-instructor-view/tutorial-groups/tutorial-groups-management/tutorial-groups-management.component.html (hunk 10 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/course/tutorial-groups/tutorial-groups-instructor-view/tutorial-groups/tutorial-groups-management/tutorial-groups-management.component.html]: It works! Changed it in 404a61f9e625e0e40fb810485d800baaea28b685

File: src/main/webapp/app/overview/course-overview.component.ts (hunk 11 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/overview/course-overview.component.ts]: Done! Changed in 404a61f9e625e0e40fb810485d800baaea28b685

**Comment by stefanwaldhauser** [comment]
> Left some comments. 🙂 I don't know how helpful the side panel (number of registrations) is for students? This should probably be only on the course management page. Screenshot 2022-10-07 at 21 19 14

I added it to course management only in 6f3254971d8ac58c2941f356180e62b61307e3d4
image

File: src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java (hunk 9 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java:81]: For example on the detail page

**Review by manuelmanso — APPROVED** [review]
Approve after changes were addressed, great job!

**Review by JohannesStoehr — DISMISSED** [review]
Code Looks good to me

**Comment by stefanwaldhauser** [comment]
I now added a basic tutor view to this PR in 400cf59c466a32e9228d13e8f407100a1ff634e3 also.

File: src/main/java/de/tum/in/www1/artemis/service/AuthorizationCheckService.java [hunk 12]
```diff
@@ -483,10 +484,41 @@ public boolean isAllowedToSeeLectureUnit(@NotNull LectureUnit lectureUnit, @Null
...
+     */
+    public boolean isAllowedToSeePrivateTutorialGroupInformation(@NotNull TutorialGroup tutorialGroup, @Nullable User user) {
+        if (user == null || user.getGroups() == null) {
+            user = userRepository.getUserWithGroupsAndAuthorities();
+        }
+        if (isAdmin(user)) {
+            return true;
+        }
+        Course course = tutorialGroup.getCourse();
+        return isAtLeastEditorInCourse(course, user) || (tutorialGroup.getTeachingAssistant() != null && tutorialGroup.getTeachingAssistant().equals(user));
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/AuthorizationCheckService.java:502]: The role of editors is to help instructors to create exercises. Should they also be allowed to see these private information?

File: src/main/webapp/i18n/de/tutorialGroups.json [hunk 13]
```diff
@@ -90,7 +90,8 @@
                 "additionalInformationNotSet": "Keine zusätzlichen Informationen festgelegt"
             },
             "registeredStudents": {
-                "title": "Registrierte Studierende"
+                "title": "Registrierte Studierende",
+                "capacityWarning": "Die Kapazität der Übungsgruppe von {{ capacity }} ist erreicht! Are you sure you want to register more students?"
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/tutorialGroups.json:94]: You should stick to one language :P

File: src/test/java/de/tum/in/www1/artemis/tutorialgroups/TutorialGroupIntegrationTest.java [hunk 14]
```diff
@@ -308,4 +355,54 @@ void registerMultipleStudents_asInstructor_shouldRegisterStudents() throws Excep
...
+        assertThat(tutorialGroup.getCourse()).isEqualTo(null);
+    }
+
+    private void verifyPrivateInformationIsShown(TutorialGroup tutorialGroup, Integer numberOfRegistrations) {
+        assertThat(tutorialGroup.getRegistrations()).hasSize(numberOfRegistrations);
+        assertThat(tutorialGroup.getTeachingAssistant()).isNotNull();
+        assertThat(tutorialGroup.getCourse()).isNotNull();
+    }
+
+    private void getOneOfCoursePrivateInfoHiddenTest() throws Exception {
```
Comment (JohannesStoehr) [inline src/test/java/de/tum/in/www1/artemis/tutorialgroups/TutorialGroupIntegrationTest.java]: Maybe setOne... or createOne... instead?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Left some comments regarding the new commit:

File: src/main/java/de/tum/in/www1/artemis/service/AuthorizationCheckService.java (hunk 12 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/AuthorizationCheckService.java]: I discussed this with @krusche and we decided that only instructors and the tutor should be able to see this and register / deregister students. So I implemented this in 677a9806c5f068adfdf9d2689df4006b3eb09206

File: src/main/webapp/i18n/de/tutorialGroups.json (hunk 13 above)
Comment (stefanwaldhauser) [reply src/main/webapp/i18n/de/tutorialGroups.json]: Haha :D Good catch! Removed english part in f25f24093da371c004de88e8a89d4984b33675b7

File: src/test/java/de/tum/in/www1/artemis/tutorialgroups/TutorialGroupIntegrationTest.java (hunk 14 above)
Comment (stefanwaldhauser) [reply src/test/java/de/tum/in/www1/artemis/tutorialgroups/TutorialGroupIntegrationTest.java]: Just named it now: oneOfCoursePrivateInfoHiddenTest. I hope thats ok?

**Review by JohannesStoehr — DISMISSED** [review]
Thanks for incorporating my suggestions. Code looks good to me now

File: src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java (hunk 9 above)
Comment (ge65cer) [reply src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java:85]: I understand your reasoning and the positive sides, but I don't think we have done this so far. There are also some downsides, like cluttering the models with these attributes. Thus, I'd be a design change and more reviewers should add their opinion on this. 🙂 
@JohannesStoehr @manuelmanso What do you say?

**Review by ge65cer — COMMENTED** [review]
Tested on TS1, looks good! Left some comments below while skimming the code. 🙂 

https://github.com/ls1intum/Artemis/pull/5718#discussion_r990441230

**Review by b-fein — COMMENTED** [review]
One comment about the translations. I did not review the remaining code in detail.

File: src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java [hunk 15]
```diff
@@ -62,6 +62,28 @@ public class TutorialGroup extends DomainObject {
...
 
+    // ==== Transient fields ====
+
+    @Transient
+    @JsonSerialize
+    private Boolean isUserRegistered;
+
+    @Transient
+    @JsonSerialize
+    private Boolean isUserTutor;
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java:79]: Added javadoc in 2592d40592be1053450720bb8edab3de2b487276

File: src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java [hunk 16]
```diff
@@ -62,6 +62,28 @@ public class TutorialGroup extends DomainObject {
     @JsonIgnoreProperties(value = "tutorialGroup", allowSetters = true)
     private Set<TutorialGroupRegistration> registrations = new HashSet<>();
 
+    // ==== Transient fields ====
+
+    @Transient
+    @JsonSerialize
+    private Boolean isUserRegistered;
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java:72]: Added javadoc in 2592d40592be1053450720bb8edab3de2b487276

File: src/main/webapp/app/course/manage/course-management.route.ts [hunk 17]
```diff
@@ -81,6 +76,13 @@ export const courseManagementState: Routes = [
         path: ':courseId/exams/:examId/plagiarism-cases',
         loadChildren: () => import('../plagiarism-cases/instructor-view/plagiarism-cases-instructor-view.module').then((m) => m.ArtemisPlagiarismCasesInstructorViewModule),
     },
+    {
+        path: ':courseId/tutorial-groups-management',
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/webapp/app/course/manage/course-management.route.ts]: Changed route in debae204182bc4dd986cb5e48fc7b667ba319f78

File: src/main/webapp/app/overview/course-overview.component.ts [hunk 18]
```diff
@@ -192,12 +200,17 @@ export class CourseOverviewComponent implements OnInit, OnDestroy, AfterViewInit
...
 
-    loadLearningGoals() {
-        forkJoin([this.learningGoalService.getAllForCourse(this.courseId), this.learningGoalService.getAllPrerequisitesForCourse(this.courseId)]).subscribe({
-            next: ([learningGoals, prerequisites]) => {
+    loadLearningGoalsAndTutorialGroups() {
+        forkJoin([
+            this.learningGoalService.getAllForCourse(this.courseId),
+            this.learningGoalService.getAllPrerequisitesForCourse(this.courseId),
+            this.tutorialGroupService.getAllForCourse(this.courseId),
+        ]).subscribe({
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/webapp/app/overview/course-overview.component.ts:208]: Good idea. But tbh the effort is not worth it imo to do this in this PR.

File: src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java (hunk 9 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java:100]: The other clean option would be a DTO but we were in the past hesitant to introduce too many DTOS.

**Review by JohannesStoehr — APPROVED** [review]
Code looks good after code reviews

**Review by JohannesStoehr — APPROVED** [review]
Merge and deletion of unused parameter look ok.
Re-approve

**Review by JohannesStoehr — APPROVED** [review]
Client test fix looks good.
Code-Reapprove

**Review by ge65cer — APPROVED** [review]
Checked the changes since my last review. Approved!
